### PR TITLE
Polymorphic multidimensional regional result writers

### PR DESCRIPTION
**Pull request by abyrd**
**Note that the base of this PR is existing PR #678, so we may choose to include it in the upcoming release testing.**
   
This is a pretty heavy refactor of the CSV writers that also exposes some latent features. All CSV writers are expressed as sharing a common interface, so independent of which ones are present we can take the same actions on all of them in loops. For example:

```Java
            if (job.templateTask.recordTimes) {
                csvResultWriters.add(new TimeCsvResultWriter(job.templateTask, outputBucket, fileStorage));
            }

            if (job.templateTask.includePathResults) {
                csvResultWriters.add(new PathCsvResultWriter(job.templateTask, outputBucket, fileStorage));
            }
// ...
            for (CsvResultWriter writer : csvResultWriters) {
                regionalAnalysis.addCsvStoragePath(writer.resultType());
            }
// ...
           if (csvResultWriters != null) {
                for (CsvResultWriter csvWriter : csvResultWriters) {
                    csvWriter.writeOneWorkResult(workResult);
                }
            }
// ...
           for (CsvResultWriter csvWriter : csvResultWriters) {
                csvWriter.finish();
            }
```

I believe this will improve the long term maintainability and comprehension of this code. I've also enabled writing out percentile, cutoff, and even destination pointset identifiers where relevant. So we can now run CSV regional analyses with multiple dimensions.

### CODE REVIEW

File: src/main/java/com/conveyal/analysis/results/CsvResultWriter.java [hunk 1]
```diff
@@ -2,76 +2,102 @@
...
         csvWriter.close();
-        // Note that stored file undergo gzip compression in super.finish(), but be stored with a .csv extension.
-        // When this file is downloaded from the UI, the browser will decompress, yielding a logically named .csv file.
-        // Downloads through another channel (e.g. aws s3 cp), will need to be decompressed manually.
         super.finish(this.fileName);
     }
 
     /**
-     * Write a single row into the CSV file.
+     * Write all rows for a single regional work result (single origin) into the CSV file.
```
Comment (ansoncfit) [inline src/main/java/com/conveyal/analysis/results/CsvResultWriter.java:88]: Good correction - the CSV is sometimes expected to have multiple rows per origin (e.g. paths)

File: src/main/java/com/conveyal/analysis/results/MultiOriginAssembler.java [hunk 2]
```diff
@@ -260,55 +218,27 @@ private synchronized void finish() {
...
-                    ArrayList<String[]> pathsIterations = pathsToPoints[d];
-                    for (String[] iterationDetails : pathsIterations) {
-                        String originId = originPointSet.getId(workResult.taskId);
-                        String destinationId = destinationId(workResult.taskId, d);
-                        checkDimension(workResult, "columns", iterationDetails.length, PathResult.DATA_COLUMNS.length);
-                        pathCsvWriter.writeOneRow(originId, destinationId, iterationDetails);
-                    }
+            if (csvResultWriters != null) {
+                for (CsvResultWriter csvWriter : csvResultWriters) {
+                    csvWriter.writeOneWorkResult(workResult);
```
Comment (ansoncfit) [inline src/main/java/com/conveyal/analysis/results/MultiOriginAssembler.java]: Much cleaner, nice.

File: src/main/java/com/conveyal/analysis/results/TimeCsvResultWriter.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        // Can nested iteration be made identical for all CSV writer subclasses, factoring it out to the superclass?
+        String originId = task.originPointSet.getId(workResult.taskId);
+        List<String[]> rows = new ArrayList<>();
+        // Only one destination pointset for now.
+        for (int p = 0; p < task.percentiles.length; p++) {
+            int[] percentileResult = workResult.travelTimeValues[p];
+            for (int d = 0; d < percentileResult.length; d++) {
+                int travelTime = percentileResult[d];
+                if (travelTime == Integer.MAX_VALUE) {
+                    travelTime = -1; // Replace 2^31 with less arcane value for end user export.
```
Comment (ansoncfit) [inline src/main/java/com/conveyal/analysis/results/TimeCsvResultWriter.java:57]: 👍

File: src/main/java/com/conveyal/analysis/results/AccessCsvResultWriter.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        for (int d = 0; d < task.destinationPointSetKeys.length; d++) {
+            int[][] percentilesForDestPointset = workResult.accessibilityValues[d];
+            for (int p = 0; p < task.percentiles.length; p++) {
+                int[] cutoffsForPercentile = percentilesForDestPointset[p];
+                for (int c = 0; c < task.cutoffsMinutes.length; c++) {
+                    rows.add(new String[] {
+                            originId,
+                            task.destinationPointSetKeys[d],
+                            Integer.toString(task.percentiles[p]),
+                            Integer.toString(task.cutoffsMinutes[c]),
```
Comment (ansoncfit) [inline src/main/java/com/conveyal/analysis/results/AccessCsvResultWriter.java:56]: Great to now handle multiple percentiles and cutoffs for CSV results.

**Review by ansoncfit — APPROVED** [review]
These refactors are helpful for legibility.

One refinement to consider: path and travel time results are clipped to the maximum of `task.cutoffsMinutes`. The UI default has a maximum of 60 minutes, and it may be confusing if results indicate destinations more than 1 hour (but less than 2 hours) away from the origin are unreachable. Possible ways to address this concern include adding 120 to the default `cutoffsMinutes` in the UI when `recordTimes` or `recordPaths` is true, or making the change in R5 (e.g. in `handleOneRegionalTask`). Thoughts @trevorgerhardt?

**Comment by ansoncfit** [comment]
Oops, I submitted my review just after the latest changes came in.

**Comment by abyrd** [comment]
I have done quite a few regional analyses with these changes over the last few hours and they seem to be working well, but if everyone's interested in getting these into the release, please continue to put them through their paces.

File: src/main/java/com/conveyal/analysis/results/MultiGridResultWriter.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ */
+public class MultiGridResultWriter implements RegionalResultWriter {
+
+    private final RegionalAnalysis regionalAnalysis;
+
+    private final RegionalTask task;
+
+    /**
+     * We create one GridResultWriter for each destination pointset and percentile.
+     * Each of those output files contains data for all travel time cutoffs at each origin.
```
Comment (ansoncfit) [inline src/main/java/com/conveyal/analysis/results/MultiGridResultWriter.java:20]:
```suggestion
     * Each of those output files contains data for all requested travel time cutoffs at each origin.
```

**Review by ansoncfit — APPROVED** [review]

**Pull request merged**